SDWebImage/FLAnimatedImage/FLAnimatedImageView+WebCache.m

```diff
@@ -139,27 +145,36 @@ - (void)sd_setImageWithURL:(nullable NSURL *)url
                                strongSelf.animatedImage = nil;
                                return;
```

Comment:
This will reach a un-balanced of `dispatch_group_enter && dispatch_group_leave`. Because for the placeholder, we also call the setImageBlock, with a `dispatch_group_enter`. If we return it without call `dispatch_group_leave`, it will crash

Reply:
@dreampiggy Good catch. 👍 

> Remove unnecessary dispatch_group_enter/leave pair, only apply to GIF data.. This is wrong. Because we always call dispatch_group_enter before setImage, for example, the placeholderImage. If the user provide a placeholderImage with GIF (Bind it one FLAnimatedImage), this change will cause a un-balanced of group enter && leave. So we must always call dispatch_group_leave in the setImageBlock.

I updated the code, I added a parameter `isPlaceholder` in `setImage` block, and I think we need a detailed discussion, the reason I add this parameter is we should know wether it's placeholder of not, e.x. for your comment above, if we have a placeholder image, it's GIF, and not have `associatedAnimatedImage`, that is, placeholder would go to `Step 3`, then we search the disk data by url, let's assume it exist, then we use this image data to create `FLAnimatedImage`, and set to `ImageView`. You can see, this would leads to `FLAnimatedImage` assign twice, one is in placeholder phase, second is when we complete load image in `loadImageWithURL:....`.

```
                           // Hack, mark we need should use dispatch group notify for completedBlock
                            objc_setAssociatedObject(group, &SDWebImageInternalSetImageGroupKey, @(YES), OBJC_ASSOCIATION_RETAIN_NONATOMIC);
                            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
                                // Step 3. Check if data exist or query disk cache
                                __block NSData *gifData = imageData;
                                if (!gifData) {
                                    NSString *key = [[SDWebImageManager sharedManager] cacheKeyForURL:url];
                                    gifData = [[SDImageCache sharedImageCache] diskImageDataForKey:key];
```

Reply:
? Why does placeholder to be considered GIF but not conatains `associatedAnimatedImage`?

`if we have a placeholder image, it's GIF, and not have associatedAnimatedImage` this case can not happen

Reply:
@dreampiggy `BOOL isGIF = (image.sd_imageFormat == SDImageFormatGIF || [NSData sd_imageFormatForImageData:imageData] == SDImageFormatGIF);` , placeholder is just a image created using `animatedImageWithImages`.

Reply:
@dreampiggy Emm, TBO, I don't know why you think I think too much. It's definitely a bug, not about performance. This case is based on your assumption.

Reply:
Is this is a rare case? I have a animated placeholder not use `FLAnimatedImage`, and the image data I request is cached in disk.

Reply:
@zhongwuzw See that `sd_imageFormat`, it's not a `animatedImageWithImages`. `_UIAnimatedImage` are not GIF image....This is what we've  changed in the 4.0, many version was passed.

You must bind the `FLAnimatedImage` on the `UIImage` instance, or you have a image data, which File Signature can be detected with GIF, or it will not treated as `GIF`. As placeholder, `imageData` is always nil. See that code with comments..

Reply:
@dreampiggy Please see where we set `animatedImage.sd_imageFormat = SDImageFormatGIF;` in the code.

```
// SDWebImageGIFCoder.m
        animatedImage = [SDWebImageCoderHelper animatedImageWithFrames:frames];
        animatedImage.sd_imageLoopCount = loopCount;
        animatedImage.sd_imageFormat = SDImageFormatGIF;
```

If you still get what I want to say. Your `PR` #2393 
>  For example, when the normal UIImageView query one GIF url, it does not create and bind a FLAnimatedImage instance. so that when another FLAnimatedImageView which query the same url, will not get a FLAnimatedImage associated image. And then it does not show animation. 

Now I think you know we can have a placeholder image, `animatedImage.sd_imageFormat == SDImageFormatGIF` but not have `associatedAnimatedImage`. Yes, as you said, this situation, `imageData` is always nil, so it would query the disk by url request, then the bug produce.

Reply:
....emmm, I still can not get the point, how will this "placeholder image is GIF, but it does not contains an associated FLAnimagedImage" happens ? The placeholder is provided by user, but not anything that from our decoder

Reply:
I can figure one point: There is no difference for a image to set, during setImageBlock. It should never consider what the image is from, but only the image to be set. The placeholder process should always be equal to the final image, or our many of feature badic assumtation are breaked. For example, view transition.

Reply:
> ....emmm, I still can not get the point, how will this "placeholder image is GIF, but it does not contains an associated FLAnimagedImage" happens ? The placeholder is provided by user, but not anything that from our decoder

case one, from bundle:
```
    NSURL *url = [[NSBundle mainBundle] URLForResource:@"loading" withExtension:@"gif"];
    NSData *data = [NSData dataWithContentsOfFile:url];
    UIImage *image = [UIImage sd_animatedGIFWithData:data];
```

case two, from server:
1. Use `SDWebImageDownloader`.
2. Get image from `UIImageView+WebCache` method.
3. .......


> I can figure one point: There is no difference for a image to set, during setImageBlock. It should never consider what the image is from, but only the image to be set. The placeholder process should always be equal to the final image, or some of feature basic assumtation are broken. For example, cell reusing (rely image to be refrshed multiple times during same runloop), view transition (rely previous placeholder and current image to be set differently)

Most cases, it's true, I don't never say we should always figure out it's placeholder or not, because for static image, `setImage` would not consider any about this. I just provide a new `API` that user can get the current image wether or not is placeholder when `setImage`. This case is used case like `GIF`, the bug I already comment [there](https://github.com/rs/SDWebImage/pull/2404#discussion_r210195749).

Reply:
Another option can temporary fix his issue, if we don't want to add placeholder parameter, remove `image.sd_imageFormat == SDImageFormatGIF` in `BOOL isGIF = (image.sd_imageFormat == SDImageFormatGIF || [NSData sd_imageFormatForImageData:imageData] == SDImageFormatGIF);`, only we have `imageData`, we can go on.

Reply:
I get your point.

```objective-c
UIImage *placeholder = [UIImage sd_animatedGIFWithData:data];
```

Actually, this usage is wrong. Because for `FLAnimatedImageView`, you should always use `FLAnimatedImage` for GIF representation, but not the thing like the code above, to create a `_UIAnimatedImage` which represent the GIF image. Mix them will cause many potential issue.

Another thing, why here we query the `diskCache` for the url, even that we does not have `imageData`. It's because of a rare case that one user face: Sometime that memory cached image hit, but the retuend `image` instance does not get associated `FLAnimatedImage`. It's may because that they previouslly query one GIF url, on the normal `UIImageView`, so the cached GIF does not contains the `FLAnimatedImage` (But actually it's quite rare, and we don't find the root case beside this potential reason).

I think this case it's not good to solve, so the previous version add a backup to query disk cache. But from your idea, I think it's not a good idea. We should also revert this query disk cache behavior, only continue when we have `nonnull imageData`.

If we remove that query disk cache behavior, so this problem solved, because placeholder does not contains `imageData`, even it's a GIF representation. So we can earily return.

See the related changes in #2393

Reply:
> if we remove that query disk cache behavior, so this problem solved, because placeholder does not contains imageData

Nope, one bug produce another bug, two situations can make `imageData` is nil, one is set placeholder, another is memory cache hit. If we set `sd_cacheFLAnimatedImage` of `FLAnimatedImageView` to `NO`, next time when we load `FLAnimatedImage`, e.x. cell reuse, memory cache would hit, because we store the `UIImage` in the memory cache, but not have `imageData`, so `FLAnimatedImage` would invalid. The image becomes static image.

So IMO we should definitely add some args in `setImage` block, two args currently we have can not satisfy needs.

So let's talk about which args we can add. I'll handle these things carefully .
@dreampiggy @bpoplauschi @mythodeia

Reply:
Emm...The 4.x seems `sd_cacheFLAnimatedImage` is optional. But 5.x, this is built-in behavior that can not disable it.

For this issue, so you means you need to check whether it's `placeholder` ? But still need to query disk cache ?

If so, I can just update that PR to introduce this check, without any public API changed.